lib/ui/window/platform_message_response_dart.cc

```diff
@@ -42,8 +48,15 @@ void PlatformMessageResponseDart::Complete(std::unique_ptr<fml::Mapping> data) {
         }
         tonic::DartState::Scope scope(dart_state);
 
-        Dart_Handle byte_buffer =
-            tonic::DartByteData::Create(data->GetMapping(), data->GetSize());
+        void* mapping = const_cast<uint8_t*>(data->GetMapping());
+        size_t data_size = data->GetSize();
+        Dart_Handle byte_buffer = Dart_NewExternalTypedDataWithFinalizer(
+            /*type=*/Dart_TypedData_kByteData,
+            /*data=*/mapping,
+            /*length=*/data_size,
+            /*peer=*/data.release(),
+            /*external_allocation_size=*/data_size,
+            /*callback=*/MappingFinalizer);
```

Comment:
Positive change. But, the previous version of the response was a copy that the Dart code could potentially modify. Now, modifications will be potentially fatal (if possible at all when using an external typed data, I'm not sure).

I suspect not too many use cases modify the response buffer, but I'm not sure. Does this require a migration?

Reply:
cc @dnfield @jonahwilliams who might know more about the mutability of these buffers.

Reply:
I think we had looked into this earlier and concluded we couldn't use this API because: we currently expose this buffers as mutable. 

If we exposed them as immutable (besides being a breaking change), Dart will choke on them if they end up getting fed back into an engine API:  https://github.com/dart-lang/sdk/issues/42785

Reply:
We need some kind of support from dart for this use case - immutable data that doesn't incur runtime penalties related to using `UnmodifiableTypedData`. /cc @lhrn - I think we've talked about this before

Reply:
Why is it a problem if they are mutated?  Since we've migrated the data buffers to unique pointers we can guarantee that Dart is the only one that has access to the data.

Reply:
I was able to run 'customer: money' with no problems with this change, but framework tests do fail:
```
aaclarke-macbookpro2:flutter aaclarke$ flutter --local-engine-src-path ~/dev/engine/src --local-engine=host_debug_unopt test
05:12 +6331 ~15: /Users/aaclarke/dev/flutter/packages/flutter/test/material/tab_bar_theme_test.dart: Tab bar theme - beveled rect indicator
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following _Exception was thrown while running async test code:
Exception: Invalid image data

When the exception was thrown, this was the stack:
#0      _futurize (dart:ui/painting.dart:5779:5)
#1      ImageDescriptor.encoded (dart:ui/painting.dart:5634:12)
#2      instantiateImageCodecFromBuffer (dart:ui/painting.dart:2098:60)
#3      instantiateImageCodec (dart:ui/painting.dart:2056:10)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)
════════════════════════════════════════════════════════════════════════════════════════════════════
05:12 +6331 ~15 -1: /Users/aaclarke/dev/flutter/packages/flutter/test/material/tab_bar_theme_test.dart: Tab bar theme - beveled rect indicator [E]
  Test failed. See exception logs above.
  The test description was: Tab bar theme - beveled rect indicator
```

I'll see if I can address this, dropping back into a Draft.  Worst case scenario we can selectively perform a copy for assets.  This copy is not required for platform channels.

Reply:
(I can't reproduce this failure again now, investigating to see if it is non-deterministic)

Reply:
I'm unable to prove if that failure is actually a bug.  It's hard when I can't reproduce it.  If we are supplying a mmaped file to dart where previously we were supplying a copy, editing the file on disk could create a problem where the ui thread would encounter a file that changed on disk underneath it.  It may be purely academic if in practice we are always passing mmaped files that are never modified on disk.  Also, that wouldn't be a new bug if the modifier is not the ui thread (this change just makes it more likely).

Reply:
> Why is it a problem if they are mutated? Since we've migrated the data buffers to unique pointers we can guarantee that Dart is the only one that has access to the data.


I was actually thinking about what happens if pages marked readonly are written to by Dart. That would be a segmentation fault. But you're also right about flutter::APKAssetMappings which are backed by [AAssets that are not thread safe](https://developer.android.com/ndk/reference/group/asset#group___asset_1ga5630b1f1aa5cd363303018cb2f12f95c).

I am positive this is a contrived use case but it ought to be easy to write a test for this in the shell_unittests harness. I am actually curious if we can cause a segfault this way.